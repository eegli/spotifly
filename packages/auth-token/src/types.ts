export type SpotifyTokenResponse = {
  access_token: string;
  token_type: 'Bearer';
  expires_in: number;
  refresh_token: string;
  scope: string;
  date_obtained: string;
};
