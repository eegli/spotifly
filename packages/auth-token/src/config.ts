import { Parser } from '@eegli/tinyparse';
import log from '@spotifly/utils/log';

export const options = new Parser()
  .setMeta({
    command: 'spotifly auth',
    help: {
      command: 'help',
      longFlag: '--help',
      shortFlag: '-h',
    },
  })
  .option('clientId', {
    required: true,
    longFlag: '--client-id',
    defaultValue: '',
    description: 'The client id of your Spotify application',
  })
  .option('clientSecret', {
    required: true,
    longFlag: '--client-secret',
    defaultValue: '',
    description: 'The client secret of your Spotify application',
  })
  .option('scopes', {
    longFlag: '--scopes',
    shortFlag: '-s',
    defaultValue: 'user-read-email',
    description:
      'Spotify authorization scopes. Multiple scopes need to be separated by a space',
  })
  .option('port', {
    longFlag: '--port',
    shortFlag: '-p',
    defaultValue: 3000,
    description: 'Port for the localhost redirect URL',
  })
  .option('outDir', {
    longFlag: '--out-dir',
    shortFlag: '-o',
    defaultValue: '',
    description: 'Custom relative output directory',
  })
  .option('fileName', {
    longFlag: '--file-name',
    shortFlag: '-f',
    defaultValue: 'spotify-token',
    description: 'Custom name for the output JSON file',
  })
  .option('noEmit', {
    longFlag: '--no-emit',
    defaultValue: false,
    description: 'When set to true, does not save the token to the file system',
  })
  .onError(({ error, usage }) => {
    log.error('Error: ' + error.message);
    log.log(usage);
  });

export type Options = typeof options;
