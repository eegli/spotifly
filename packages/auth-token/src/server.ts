import http from 'http';

export const localhostUrl = async (port: number): Promise<string> => {
  return new Promise((resolve, reject) => {
    const server = http
      .createServer((req, res) => {
        res.writeHead(200, { 'Content-Type': 'text/plain' });
        res.end('You can now close this window');
        res.once('finish', () => {
          server.close();
          if (req.url) {
            resolve(req.url.slice(1));
          }
          reject("Couldn't get code or state");
        });
      })
      .listen(port);
  });
};
