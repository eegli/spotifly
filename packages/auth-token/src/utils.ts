export const randomState = () => Math.random().toString(36).slice(2);
