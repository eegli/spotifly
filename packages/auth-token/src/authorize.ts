import { writeJSON } from '@spotifly/utils';
import fetch from 'node-fetch';
import { defaultConfig } from './config';
import { localhostUrl } from './server';
import type { Options, SpotifyTokenResponse } from './types';
import { id } from './utils';

export const authorize = async (
  options: Options,
): Promise<SpotifyTokenResponse> => {
  const config = { ...defaultConfig, ...options };
  const redirectUri = `http://localhost:${config.port}`;
  const state = id();

  const spotifyUrl =
    'https://accounts.spotify.com/authorize?' +
    new URLSearchParams({
      response_type: 'code',
      show_dialog: 'true',
      state,
      client_id: config.clientId,
      redirect_uri: redirectUri,
      scope: config.scopes,
    }).toString();

  console.info('Please click the link to login to Spotify in the browser\n');
  console.info(spotifyUrl + '\n');

  const authUrl = await localhostUrl(config.port);
  const params = new URLSearchParams(authUrl);
  const receivedCode = params.get('code');
  const receivedState = params.get('state');

  if (receivedState !== state) {
    throw new Error('Received and original state do not match');
  }

  if (!receivedCode) {
    throw new Error('No code received');
  }

  console.info('Login successfull! Cleaning up...\n');

  const tokenRequestBody = new URLSearchParams({
    grant_type: 'authorization_code',
    code: receivedCode,
    redirect_uri: redirectUri,
  });

  const token: SpotifyTokenResponse = await fetch(
    'https://accounts.spotify.com/api/token',
    {
      method: 'POST',
      headers: {
        'Content-type': 'application/x-www-form-urlencoded',
        Authorization:
          'Basic ' +
          Buffer.from(config.clientId + ':' + config.clientSecret).toString(
            'base64',
          ),
      },
      body: tokenRequestBody.toString(),
    },
  ).then(res => res.json());

  token.date_obtained = new Date().toUTCString();

  if (!config.noEmit) {
    const outDir = await writeJSON({
      path: config.outDir,
      fileName: config.fileName,
      data: token,
    });
    console.info(`Success! Saved Spotify access token to "${outDir}"`);
  }

  return token;
};
