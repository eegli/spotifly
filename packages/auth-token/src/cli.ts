import { authorize } from './authorize';
import { options } from './config';

export const parser = options.defaultHandler(authorize);
