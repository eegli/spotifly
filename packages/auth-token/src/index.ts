import { WithOptions } from '@eegli/tinyparse';
import { authorize as internalAuthorize } from './authorize';
import { parser } from './cli';
import type { options } from './config';

type InternalOptions = WithOptions<typeof options>['options'];
type OptionalOptions = Partial<InternalOptions>;
type RequiredOptions = Pick<InternalOptions, 'clientId' | 'clientSecret'>;

export type Options = OptionalOptions & RequiredOptions;

export async function authorize(authOptions: Options) {
  return internalAuthorize({ options: { ...parser.options, ...authOptions } });
}
