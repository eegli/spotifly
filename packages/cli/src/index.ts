import { run } from './cli';

(async () => {
  await run();
})();
