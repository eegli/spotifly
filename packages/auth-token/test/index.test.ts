import * as fs from '@spotifly/utils/fs';
import log from '@spotifly/utils/log';
import { Options, authorize } from '../src/index';
import * as server from '../src/server';
import { SpotifyTokenResponse } from '../src/types';
import * as utils from '../src/utils';

const fakeDate = new Date('1996-04-20T22:00:00.000Z');
const returnedToken: SpotifyTokenResponse = {
  access_token: 'access_token',
  refresh_token: 'refresh_token',
  expires_in: 3600,
  scope: 'user-read-private user-read-email',
  token_type: 'Bearer',
  date_obtained: fakeDate.toUTCString(),
};
const testState = 'ou8n1fu8n1';

jest.useFakeTimers().setSystemTime(fakeDate);

jest.mock(
  'node-fetch',
  () => () =>
    Promise.resolve({
      json: () => Promise.resolve(returnedToken),
    }),
);

jest.spyOn(utils, 'randomState').mockReturnValue(testState);

const localhostSpy = jest
  .spyOn(server, 'localhostUrl')
  .mockResolvedValue(`?code=123&state=${testState}`);

const logInfoSpy = jest.spyOn(log, 'info').mockImplementation(() => {});
const logLogSpy = jest.spyOn(log, 'log').mockImplementation(() => {});
const logErrorSpy = jest.spyOn(log, 'error').mockImplementation(() => {});

const writeSpy = jest
  .spyOn(fs, 'writeJSON')
  .mockImplementation(opts => Promise.resolve(opts.path));

afterEach(() => {
  jest.clearAllMocks();
});

const params: Options = {
  clientId: 'cid',
  clientSecret: 'secret',
};

describe('Authorize via package', () => {
  it("fails if states don't match", async () => {
    localhostSpy.mockResolvedValueOnce(`?code=123&state=${testState + 'XXX'}`);
    const userToken = await authorize(params);
    expect(userToken).toBeUndefined();
    expect(logErrorSpy).toHaveBeenCalledWith(
      'Error: Received and original state do not match',
    );
  });

  it('fails if no code is received', async () => {
    localhostSpy.mockResolvedValueOnce(`?state=${testState}`);
    const userToken = await authorize(params);
    expect(userToken).toBeUndefined();
    expect(logErrorSpy).toHaveBeenCalledWith(
      'Error: No code received from Spotify, did you cancel the login?',
    );
  });

  it('works with emit mode', async () => {
    const userToken = await authorize({
      ...params,
      outDir: 'some-dir',
      port: 1234,
      noEmit: false,
      scopes: 'user-read-email user-read-private',
      fileName: 'some-file',
    });
    expect(userToken).toStrictEqual(returnedToken);
    expect(logLogSpy.mock.calls).toMatchSnapshot('log');
    expect(logInfoSpy.mock.calls).toMatchSnapshot('info');
    expect(localhostSpy).toHaveBeenCalledWith(1234);
    expect(writeSpy).toHaveBeenCalledTimes(1);
  });

  it('works with no emit mode', async () => {
    const userToken = await authorize({
      ...params,
      noEmit: true,
    });
    expect(userToken).toStrictEqual(returnedToken);
    expect(logInfoSpy.mock.calls).toMatchSnapshot('info');
    expect(writeSpy).not.toHaveBeenCalled();
  });

  it('doc examples', async () => {
    {
      const token = await authorize({
        clientId: 'clientId',
        clientSecret: 'clientSecret',
        port: 3000,
        scopes: 'user-read-email user-top-read',
      });
      expect(token).toStrictEqual(returnedToken);
    }
    {
      const options: Options = {
        clientId: 's',
        clientSecret: 's',
        noEmit: true,
      };

      const token = await authorize(options);
      expect(token).toStrictEqual(returnedToken);
    }
  });
});
