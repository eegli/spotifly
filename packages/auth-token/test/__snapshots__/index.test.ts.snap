// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authorize via package works with emit mode: info 1`] = `
[
  [
    "Success! Saved Spotify access token to "some-dir"",
  ],
]
`;

exports[`Authorize via package works with emit mode: log 1`] = `
[
  [
    "Please click the link to login to Spotify in the browser
",
  ],
  [
    "https://accounts.spotify.com/authorize?response_type=code&show_dialog=true&state=ou8n1fu8n1&client_id=cid&redirect_uri=http%3A%2F%2Flocalhost%3A1234&scope=user-read-email+user-read-private
",
  ],
  [
    "Login successfull! Cleaning up...
",
  ],
]
`;

exports[`Authorize via package works with no emit mode: info 1`] = `
[
  [
    "Token:
{
  "access_token": "access_token",
  "refresh_token": "refresh_token",
  "expires_in": 3600,
  "scope": "user-read-private user-read-email",
  "token_type": "Bearer",
  "date_obtained": "Sat, 20 Apr 1996 22:00:00 GMT"
}",
  ],
]
`;
